### PR TITLE
SAP Monitor Create will also accept keyvault URL for creds

### PR DESCRIPTION
**Pull request by PakDLiu**
SAP Monitor Create will also accept keyvault URL for creds instead of clear text password. 
Workflow:
1. Create MSI
2.  Assign MSI to customer's key vault with get permissions
3. Pass MSI ID to RP along with the link to password

### CODE REVIEW

File: azext_hanaonazure/custom.py [hunk 1]
```diff
@@ -92,16 +95,69 @@ def list_sapmonitor(client):
...
+        # Password was passed in
+        monitoring_details.update({"hanaDbPassword": hana_db_password})
+    elif hana_db_password_key_vault_url is not None and key_vault_id is not None:
+        # Keyvault URL was passed in
+        from ._client_factory import (_msi_client_factory, _keyvault_client_factory)
+        from azure.mgmt.keyvault.v2018_02_14.models import (VaultAccessPolicyProperties, AccessPolicyEntry, Permissions, SecretPermissions)
+
+        # Create MSI
+        msi_client = _msi_client_factory(cmd.cli_ctx)
+        msi = msi_client.user_assigned_identities.create_or_update(resource_group_name, 'sapmon-msi', region)
```
Comment (tniek) [inline azext_hanaonazure/custom.py]: I'm mindful of the fact that we currently may not have access to SapMonId yet at this stage, but as per new [naming convention](https://microsoft.sharepoint.com/teams/AzureDedicatedEngineeringTeam/_layouts/OneNote.aspx?id=%2Fteams%2FAzureDedicatedEngineeringTeam%2FSiteAssets%2FAzure%20Dedicated%20Engineering%20Team%20Notebook&wd=target%28SAP%20HANA.one%7CFB7D0654-1276-104F-AF53-BA55FC4DA39E%2FDesign%7CE2184110-0189-3940-AC33-F400641A036D%2F%29), this should be `sapmon-msi-XXXXX` (with XXXXX being the SapMonId).

File: azext_hanaonazure/custom.py [hunk 2]
```diff
@@ -92,16 +95,69 @@ def list_sapmonitor(client):
...
+
+        # Extract Key Vault information
+        match = re.search(keyvault_id_format, key_vault_id)
+        if not match:
+            raise ValueError("key_vault_id is of incorrect format. The ID should start with /subscription/")
```
Comment (tniek) [inline azext_hanaonazure/custom.py]: nit: "should start with /subscription**s**/"

**Review by tniek — COMMENTED** [review]
See my comment on the naming of the MSI created inside the managed RG.

**Review by dwj300 — DISMISSED** [review]
Can you add a VCR unit test?

File: azext_hanaonazure/custom.py (hunk 1 above)
Comment (pabowers) [reply azext_hanaonazure/custom.py]: we haven't created the sapmonID at this point tho, right?

Comment (PakDLiu) [reply azext_hanaonazure/custom.py]: we haven't, I am looking into how to generate the sapmonID in python

**Review by jovillas — APPROVED** [review]
LGTM

File: HISTORY.rst [hunk 3]
```diff
@@ -3,6 +3,11 @@
 Release History
 ===============
 
+0.5.2 (2019-07-24)
+++++++++++++++++++
+
+- sapmonitor create will now accept key vault as well
```
Comment (Caiyeon) [inline HISTORY.rst]: This sentence doesn't make sense on its own. I would suggest being more clear (e.g. accept key vault url for credentials)

File: azext_hanaonazure/custom.py [hunk 4]
```diff
@@ -92,17 +96,81 @@ def list_sapmonitor(client):
...
+        })
+    else:
+        raise ValueError("Either --hana-db-password or both --hana-db-password-key-vault-url and --key-vault-id.")
+
     return client.create(resource_group_name, monitor_name, monitoring_details)
 
 def delete_sapmonitor(client, resource_group_name, monitor_name):
     return client.delete(resource_group_name, monitor_name)
+
+def fnv32a(str):
```
Comment (Caiyeon) [inline azext_hanaonazure/custom.py]: The purpose of this function isn't immediately obvious from the name. Can we add a comment describing it? (purpose, input, output)

File: azext_hanaonazure/custom.py [hunk 5]
```diff
@@ -92,17 +96,81 @@ def list_sapmonitor(client):
...
+    hval = 0x811c9dc5
+    fnv_32_prime = 0x01000193
+    uint32_max = 2 ** 32
+    for s in str:
+        hval = hval ^ ord(s)
+        hval = (hval * fnv_32_prime) % uint32_max
+    return hval
```
Comment (Caiyeon) [inline azext_hanaonazure/custom.py]: nit: new line at end

**Review by Caiyeon — APPROVED** [review]

File: azext_hanaonazure/custom.py [hunk 6]
```diff
@@ -92,17 +96,81 @@ def list_sapmonitor(client):
...
+        monitoring_details.update({"hanaDbPassword": hana_db_password})
+    elif hana_db_password_key_vault_url is not None and key_vault_id is not None:
+        # Keyvault URL was passed in
+        from ._client_factory import (_msi_client_factory, _keyvault_client_factory)
+        from azure.mgmt.keyvault.v2018_02_14.models import (VaultAccessPolicyProperties, AccessPolicyEntry, Permissions, SecretPermissions)
+
+        # Create MSI
+        msi_client = _msi_client_factory(cmd.cli_ctx)
+        arm_resource_id = arm_resource_id_format.format(msi_client.config.subscription_id, resource_group_name, monitor_name)
+        sapmon_id = 'sapmon{}'.format(fnv32a(arm_resource_id))
```
Comment (tniek) [inline azext_hanaonazure/custom.py]: Should be: `sapmon_id = 'sapmon-msi-{}'.format(fnv32a(arm_resource_id))`
As per [naming convention](https://microsoft.sharepoint.com/teams/AzureDedicatedEngineeringTeam/_layouts/OneNote.aspx?id=%2Fteams%2FAzureDedicatedEngineeringTeam%2FSiteAssets%2FAzure%20Dedicated%20Engineering%20Team%20Notebook&wd=target%28SAP%20HANA.one%7CFB7D0654-1276-104F-AF53-BA55FC4DA39E%2FDesign%7CE2184110-0189-3940-AC33-F400641A036D%2F%29)

**Review by tniek — COMMENTED** [review]
Name of generated MSI needs to be fixed - otherwise looks good

**Comment by AprilZhaoWY** [comment]
> Can you add a VCR unit test?
@dwj300 do we have documentation on how to create the VCRs?

File: azext_hanaonazure/_params.py [hunk 7]
```diff
@@ -37,6 +37,10 @@ def load_arguments(self, _):
                    '--hana-db-username'], help="Username for the HANA database to login to for monitoring")
         c.argument('hana_db_password', options_list=[
                    '--hana-db-password', '--hdbpw'], help="Password for the HANA database to login for monitoring")
+        c.argument('hana_db_password_key_vault_url', options_list=[
+            '--hana-db-password-key-vault-url', '--hdpkvu'], help="URL to the KeyVault secret that contains the HANA credentials")
+        c.argument('key_vault_id', options_list=[
+            '--key-vault-id', '--kvi'], help="Key Vault ID containing that contains the HANA credentials")
```
Comment (AprilZhaoWY) [inline azext_hanaonazure/_params.py]: This sentence does not make sense to me, typo?

**Pull request merged**